### PR TITLE
Use the input buffer as outgoing payload in several syscalls

### LINKED ISSUE #1669: Allow reply_push to use input buffer
This will allow zero-copy message relaying in multiple scenarios

### PR DESCRIPTION
**Pull request by gshep**
Resolves #1669 

- `Ext::send_push_input` & `Ext::reply_push_input`
- adjust `gcore` & `gstd` accordingly
- implement syscalls in wasmi & sandbox backends
- add benchmarks for the new functions
- test case with relaying parts of an incoming message to different destinations

**Release notes**: This PR introduces a new ability to relay incoming messages. The slice of the incoming message buffer is used as a payload for the outgoing message. The new `gstd` API supports common range syntax (`..`, `0..`, `..=n`, etc) in Rust so one is able to relay the first half of the message the following way:
```rust
let len = gstd::msg::size() / 2 as usize;
gstd::msg::send_input(user_or_program_id, 0, ..=len).expect("`send_input' failed");
```

@gear-tech/dev

### CODE REVIEW

**Comment by NikVolf** [comment]
And more thoughtful examples where you send different part of the incoming payloads to different destination would be great

**Review by NikVolf — CHANGES_REQUESTED** [review]
Please review my comments

File: core-backend/sandbox/src/env.rs [hunk 1]
```diff
@@ -161,6 +161,12 @@ where
             Funcs::reservation_reply_commit,
         );
         builder.add_func("gr_reply_push", Funcs::reply_push);
+        builder.add_func("gr_rereply", Funcs::rereply);
```
Comment (NikVolf) [inline core-backend/sandbox/src/env.rs:164]: i think naming should be made more consistent

for example,`gr_reply_push` should have an analog of `gr_reply_push_input` and so on

Comment (gshep) [reply core-backend/sandbox/src/env.rs:164]: Yep, agree. What is about 'gr_relay_{send, push}`?

Comment (NikVolf) [reply core-backend/sandbox/src/env.rs:164]: Relay has too opionated meaning, it's not just about relay

File: core/src/message/context.rs [hunk 2]
```diff
@@ -313,6 +313,43 @@ impl MessageContext {
...
+            excluded_end,
+        } = range;
+
+        data.try_extend_from_slice(&self.current.payload()[offset..excluded_end])
+            .map_err(|_| Error::MaxMessageSizeExceed)?;
+
+        Ok(())
+    }
+
+    /// Check range for `send_push_input`/`reply_push_input` methods.
```
Comment (NikVolf) [inline core/src/message/context.rs:337]: Can you explain in doc what is this check about?

So far I don't see this functions checks anything but rather returns some different range

**Review by NikVolf — APPROVED** [review]
other than comment above

File: core/src/message/context.rs (hunk 2 above)
Comment (gshep) [reply core/src/message/context.rs:337]: ok

File: Cargo.lock [hunk 3]
```diff
@@ -1679,6 +1679,16 @@ dependencies = [
  "scale-info",
 ]
 
+[[package]]
```
Comment (breathx) [inline Cargo.lock:1682]: To be merged after #1784

**Pull request merged**